File: practice_in_python/multi-threading/primes.py
from threading import Thread
from typing import Final
import numpy as np
import time
import math

UPPER_BOUND: Final = 10**8
NUM_THREADS: Final = 8

def SieveEmForGood(startInd: int, endInd: int, primes: list[bool]) -> None:

    res = int(math.sqrt(endInd))
    for i in range(startInd, res, 2):
        if primes[i]:
            for j in range(i*i, endInd, i):
                primes[j] = False


if __name__ == '__main__':

    # starting points for the sieve, one for each thread
    primeNums = np.array([3, 5, 7, 11, 13, 17, 31, 41])

    # Set the initial values of the primes array
    primes = np.array([True] * (UPPER_BOUND + 1))
    primes[0], primes[1] = False, False

    threads = np.array([])

    start = time.time()

    for i in range(NUM_THREADS):
        threads = np.append(threads, Thread(target=SieveEmForGood, args=(primeNums[i], UPPER_BOUND, primes)),)

    for thread in threads:
        thread.start()

    for thread in threads:
        thread.join()

    end = time.time()

    print('{} ms'.format((end - start) * 1e3))

    results = [2]
    total = 2

    fast_append = results.append

    for i in range(3, UPPER_BOUND, 2):

        if primes[i]:
            fast_append(i)
            total += i

    print('total number of primes under 10^8: {}'.format(len(results)))
    print('total sum of all primes under 10^8: {}'.format(total))

    print('largest 10 primes: {}'.format(results[-10:]))

    print('done')










    pass

File: practice_in_dart/primes.dart
import 'dart:isolate';
import 'dart:math';
import 'dart:io';

final int UPPER_BOUND = 1e8.toInt();

List<Isolate> isolates = [];

// number of primes will also be the number of times isolate.spawn is called
var primeNums = [3, 5, 7, 11, 13, 17, 31, 41];
var primes = new List<bool>.filled(UPPER_BOUND + 1, true);

class Values {
  int startInd;
  int endInd;
  List<bool> primes;

  Values(this.startInd, this.endInd, this.primes);
}

void SieveEmForGood(Values obj) {
  int res = sqrt(obj.endInd).toInt();
  for (int i = obj.startInd; i < res; i += 2) {
    if (obj.primes[i]) {
      // sieve to the upper bound
      for (int j = i * i; j < obj.endInd; j += i) {
        // cross it out!
        // The given index is not prime!
        obj.primes[j] = false;
      }
    }
  }
}

void start() async {
  for (var i = 0; i < primeNums.length; i++) {
    isolates.add(await Isolate.spawn(
        SieveEmForGood, new Values(primeNums[i], UPPER_BOUND, primes)));
  }
}

void stop() {
  for (Isolate i in isolates) {
    if (i != null) {
      i.kill(priority: Isolate.beforeNextEvent);
      i = null;
      print('Killed');
    }
  }
}

void main(List<String> args) async
{
  // var l = [1, 2, 3];
  // try {
  //   Iterable<int> it = l.getRange(0, 4);
  //   print(it);
  // } catch (e) {
  //   print(e);
  // }

  primes[0] = primes[1] = false;

  Stopwatch s = new Stopwatch()..start();

  await start();

  // print('Press enter to exit');
  // await stdin.first;

  stop();

  print('took ${s.elapsedMilliseconds} ms');

  var results = [2];
  int sum = 2;

  for (var i = 3; i < UPPER_BOUND; i+=2) {
    if (primes[i]) {
      results.add(i);
      sum += i;
    }
  }

  int len = results.length;

  print('number of prime numbers under 10^8: ${results.length}');
  print('total sum of all prime numbers under 10^8: ${sum}');

  print('ten largest primes: ${results.getRange(len - 10, len)}');

  print('done');
  stdin.first;
}
